src/replicache.ts

```diff
@@ -126,6 +151,94 @@ export default class Replicache implements ReadTransaction {
     }
   }
 
+  /**
+   * Registers a *mutator*, which is used to make changes to the data.
+   *
+   * ## Replays
+   *
+   * Mutators run once when they are initially invoked, but they might also be
+   * *replayed* multiple times during sync. As such mutators should not modify
+   * application state directly. Also, it is important that the set of
+   * registered mutator names only grows over time. If Replicache syncs and
+   * needed mutator is not registered, it will substitute a no-op mutator, but
+   * this might be a poor user experience.
+   *
+   * ## Server application
+   *
+   * During sync, a description of each mutation is sent to the server's [batch
+   * endpoint](https://github.com/rocicorp/replicache/blob/master/README.md#step-5-upstream-sync)
+   * where it is applied. Once the mutation has been applied successfully, as
+   * indicated by the [client
+   * view](https://github.com/rocicorp/replicache/blob/master/README.md#step-2-downstream-sync)'s
+   * `lastMutationId` field, the local version of the mutation is removed. See
+   * the [design
+   * doc](https://github.com/rocicorp/replicache/blob/master/design.md) for
+   * additional details on the sync protocol.
+   *
+   * ## Transactionality
+   *
+   * Mutators are atomic: all their changes are applied together, or none are.
+   * Throwing an exception aborts the transaction. Otherwise, it is committed.
+   * As with [query] and [subscribe] all reads will see a consistent view of
+   * the cache while they run.
+   */
+  register<Return extends JsonType | void, Args extends JsonType>(
+    name: string,
+    mutatorImpl: MutatorImpl<Return, Args>,
+  ): Mutator<Return, Args> {
+    this._mutatorRegistry.set(
+      name,
+      (mutatorImpl as unknown) as MutatorImpl<JsonType, JsonType>,
```

Comment:
[help] Wow, can the compiler really not upcast to `MutatorImpl<JsonType, JsonType>` more gracefully than this?

Reply:
It is not type safe due to contra-variant (or co-variant?). Dart complains about the exact same thing.

Let me try explain why: `mutatorImpl` takes `Args` which can be stricter than `JsonType`; for example `string`. If you store that function in the `Map` and retrieve it and then call it with 'number' (which is a `JsonType`) it will blow up. So, `mutatorImpl<JsonType, string>` is not compatible with `mutatorImpl<JsonType, JsonType>` (also, the ordering of the type params should be reversed. I'll file an issue.)

In our case it is not going to call runtime errors because we only call the function with the right types.